text/3627-match-ergonomics-2024.md

```diff
@@ -0,0 +1,405 @@
+<https://users.rust-lang.org/t/cannot-deconstruct-reference-inside-match-on-reference-why/92147>,
+<https://github.com/rust-lang/rust/issues/50008>,
+<https://github.com/rust-lang/rust/issues/64586>).
+
+
+```rust
+fn foo(arg: &(String, Vec<i32>, u8)) {
+    // We want to extract `&String`, `&Vec`, and `u8` from the tuple.
+    let (s, v, u) = arg; // u is &u8, not what we wanted
+    let &(ref s, ref v, u) = arg; // we have to abandon match ergonomics entirely
+}
+```
+
+## A single `&` can strip two references
+
+When an `&` or `&mut` pattern is used in a location where there is also an
+inherited reference present, both are stripped; adding a single `&` to the
+pattern can remove two `&`s from the type of the binding.
+
+```rust
+let [a] = &[&42]; // a = &&42
+let [&a] = &[&42]; // a = 42
+```
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Match ergonomics works a little differently in edition 2024 and above.
+
+## `mut` no longer strips the inherited reference
+
+`mut` on a binding does not reset the binding mode on edition ≥ 2024.
+
+```rust
+//! Edition ≥ 2024
+let (x, mut y) = &(true, false);
+let _: (&bool, &bool) = (x, y); // instead of `(&bool, bool)`
+```
+
+## Matching against inherited references
+
+In all editions, when you match against an `&` or `&mut` reference with the type
+of its referent, you get an "inherited reference": the binding mode of
+"downstream" bindings is set to `ref` or `ref mut`.
+
+```rust
+//! All editions
+// `x` "inherits" the `&` from the scrutinee type.
+let [x] = &[42];
+let _: &u8 = x;
+```
+
+In edition 2024 and above, an `&` or `&mut` pattern can match against this
+inherited reference, consuming it. A pattern that does this has no other effect.
+
+```rust
+//! Edition ≥ 2024
+
+// `&` pattern consumes inherited `&` reference.
+let [&x] = &[42];
+let _: u8 = x;
+
+// Examples from motivation section
+
+fn foo(arg: &(String, Vec<i32>, u8)) {
+    let (s, v, &u) = arg;
+    let _: (&String, &Vec<i32>, u8) = (s, v, u);
+}
+
+let [&x] = &[&42];
+let _: &u8 = x;
+```
+
+## `&` matches against `&mut`
+
+In edition 2024 and above, `&` patterns can match against `&mut` references
+(including "inherited" references).
+
+```rust
+//! Edition ≥ 2024
+let &foo = &mut 42;
+let _: u8 = foo;
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This explanation assumes familiarity with the current match ergonomics rules,
+including the "default binding mode" terminology. Refer to [RFC 2005](./2005-match-ergonomics.md#detailed-design).
+
+## Edition 2024: `mut` does not reset binding mode to by-value
+
+In the new edition, `mut` no longer resets the binding mode to by-value.
+Therefore, it is possible to have a mutable by-reference binding. (An explicit
+syntax for this is left to a future RFC.)
+
+```rust
+//! Edition ≥ 2024
+let &[mut a] = &[42];
+a = &47;
+```
```

Comment:
I had no big trouble following the RFC until this example. Why is `a` a reference here? The reason it's a mutable binding is clear, but I see a `&` on both sides of `=` so shouldn't `a` just be a plain integer?

Also, would `let [mut a] = &[42];` have the same behavior? That's what I would probably want to write to get a reference binding that's mutable.

Reply:
Aha! That explains why I got confused 😄